File: src/0077-Combinations/0077.py
import itertools
class Solution:
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        nums = [i for i in range(1,n+1)]
        return list(itertools.combinations(nums,k))

# class Solution:
#     def combine(self, n, k):
#         """
#         :type n: int
#         :type k: int
#         :rtype: List[List[int]]
#         """
#         result = list()
#         pool = [i for i in range(1, n + 1)]
#         if k > n:
#             return
#         indices = [i for i in range(k)]
#         result.append([pool[i] for i in indices])
#         print([i for i in reversed(range(k))])
#         while True:
#             for i in reversed(range(k)):
#                 if indices[i] != i + n - k:
#                     break
#             else:
#                 return result

#             indices[i] += 1
#             for j in range(i+1, k):
#                 indices[j] = indices[j-1] + 1
#             result.append([pool[i] for i in indices])    

if __name__ == '__main__':
    n, k = 4, 2
    print(Solution().combine(n, k))

File: src/0077-Combinations/0077.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution 
{
public:
    vector<vector<int>> combine(int n, int k) 
    {
        vector<vector<int>> result;
        if (n <= 0 or k <= 0 or n < k) return result;
        vector<int> nums;
        _combine(n, k, 1, nums, result);
        return result;
    }
private:
    void _combine(int n, int k, int start, vector<int> nums, vector<vector<int>>& result)
    {
        if (k == nums.size())
        {
            result.push_back(nums);
            return;
        }

        for (int i = start; i <= n - (k - nums.size()) + 1; ++i)
        {
            nums.push_back(i);
            _combine(n, k, i + 1, nums, result);
            nums.pop_back();
        }
    }
};